### PR TITLE
Implementation of nn.CrossEntropyLossWithSoftLabels

### LINKED ISSUE #11959: [feature request] Support soft target distribution in cross entropy loss
Currently our cross entropy loss (i.e., [nn.CrossEntropyLoss](https://pytorch.org/docs/stable/nn.html#crossentropyloss)) only supports a hard target class, i.e., wanting to maximize the output (log) probability of a particular class. But in many times training w.r.t. a soft target distribution (i.e., wanting the output to match a particular distribution) is quite useful too, e.g., preventing overfitting.

## Math
Cross entropy loss operates on logits after softmax. 

Denote the input vector as `x`. Log softmax computes a vector `y` of same length as `x`, where `y_i = x_i - log( \sum_j exp(x_j) )`, representing the log likelihood of each class.

+ In the hard target case, if the target clss is `c`, the loss is simply negative log likelihood loss `-y_c`.

+ In the soft target case, let the target distribution vector be `p` (i.e., `p_i` is the target probability for predicting class `i`). The loss is the KL divergence 

  ```
  D( softmax(x) || p) = \sum_i p_i (log p_i  / softmax(x)_i) = -\sum_i p_i y_i + constant
  ```
  
  The constant is independent of `x` and thus discarded. Our loss formula is just `-\sum_i p_i y_i`. 
  
  When `p_c = 1` for some class `c`, this simplifies to the hard target class.
  

  The formula for gradient computation can be easily derived from this:
  
  ```
  
  d l / d y_i = -p_i
  
  d y_i / d x_i = 1 - exp(x_i) / \sum_j exp(x_j) = 1 - exp(y_i)
  
  # suppose k != i
  d y_k / d_x_i = -exp(x_i) / \sum_j exp(x_j) = - exp(y_i)
  
  # so
  d l / d x_i = exp(y_i) (\sum p) - p_i = exp (y_i) - p_i (= softmax(x) - p_i).
  
  ```

## Possible Implementation

Currently our cross entropy loss implementation takes in batched `x` of shape `(N, C)` and floating point dtype (`N` is the batch size and `C` is the number of classes), and a batched target class indices vector `target` of shape `(N)`, where `target[i]` is the index of the desired output class, and dtype `long` (an integral type).

Since we want it to also take in soft target distribution as target, we can allow it to also take in `target` as a target batched distribution of shape `(N, C)`, and detect whether we want soft target or hard target basing on shape and dtype.

cc @gchanan

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #11959. The implementation is a combination of `log_softmax` + `kl_div` (see https://github.com/pytorch/pytorch/issues/11959#issuecomment-854076986 for details).

Btw this specific name was chosen for a few reasons:
1. To distinguish the criterion from `SoftMarginLoss`, where "soft" has a different meaning applying to the margin (in contrast to a hard margin)
2. To fit with the "With" naming adopted by `BCELossWithLogits`
3. To fit with the "Labels" naming adopted by `MultiLabelMarginLoss` and `MultiLabelSoftMarginLoss`

This ruled out:
* `SoftCrossEntropyLoss`
* `SoftLabelCrossEntropyLoss`
* `CrossEntropyLossWithSoftTargets`
* `SoftTargetCrossEntropyLoss`
* etc.

### CODE REVIEW

**Comment by jbschlosser** [comment]
@zou3519 Are you cool with reviewing this? I think you have the most familiarity with this issue within the frontend group

**Comment by zou3519** [comment]
Yup, can do!

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -1121,6 +1125,91 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
                                ignore_index=self.ignore_index, reduction=self.reduction)
 
 
+class CrossEntropyLossWithSoftLabels(_Loss):
+    r"""This criterion combines :class:`~torch.nn.LogSoftmax` and :class:`~torch.nn.KLDivLoss` in
```
Comment (zou3519) [inline torch/nn/modules/loss.py:1129]: If anyone else is interested in reading the rendered docs: https://14108352-65600975-gh.circle-artifacts.com/0/docs/generated/torch.nn.CrossEntropyLossWithSoftLabels.html?highlight=crossentropy#torch.nn.CrossEntropyLossWithSoftLabels

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -1121,6 +1125,91 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
+    "soft labels" are required, such as when utilizing label smoothing or blended labels.
+
+    The `input` is expected to contain raw, unnormalized scores for each class and the `target`
+    should match the `input` shape and contain the target distribution over the classes.
+
+    The loss can be described as:
+
+    .. math::
+        L = l_{n,c} = y_{n,c} \cdot \left( \log y_{n,c} - \\
+        \log\left(\frac{\exp(x_{n,c})}{\sum_{i=1}^C \exp(x_{n,i})} \right) \right)
```
Comment (zou3519) [inline torch/nn/modules/loss.py:1140]: I noticed that in #11959, @SsnL dropped the constant term. In [elegy, a jax nn library](https://github.com/poets-ai/elegy/blob/master/elegy/losses/categorical_crossentropy.py), the constant term is also dropped. I haven't been able to figure out how to read the TF source code for categorial_crossentropy yet, so idk what they're doing there.

Should we drop the constant term as well? This would change the implementation of this loss function from being a composition of LogSoftmax and KLDivLoss to being something else that is (1) more efficient to compute and (2) optimizes the same way (assuming your targets don't require grad...) but (3) isn't exactly the quantity users may expect.

Comment (zou3519) [reply torch/nn/modules/loss.py:1140]: I think this is the TF source code for categorical_crossentropy and it looks like they drop the constant term as well: https://github.com/keras-team/keras/blob/0eb1954092e2110868e8ef06381cebac891e01e9/keras/backend.py#L4872

Comment (jbschlosser) [reply torch/nn/modules/loss.py:1140]: Great question- definitely true that it'd be more efficient to drop the constant from an optimization perspective. Not sure what the impact is for 3 (user experience), since the one-hot case will still generally be equivalent between soft and hard cross losses (since the constant=0 for that case).

FWIW, SsnL doesn't think the constant should be dropped by default: https://github.com/pytorch/pytorch/issues/11959#issuecomment-861880668. I wonder if it makes sense for this PR to leave as-is and possibly add a flag for dropping the constant later on?

Comment (zou3519) [reply torch/nn/modules/loss.py:1140]: Seems reasonable. Let's go ahead with this unless we get feedback from the other folks in the thread that they do want the constant dropped.

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -1121,6 +1125,91 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
 
+class CrossEntropyLossWithSoftLabels(_Loss):
+    r"""This criterion combines :class:`~torch.nn.LogSoftmax` and :class:`~torch.nn.KLDivLoss` in
+    one single class. It is useful when training a classification problem with `C` classes when
+    "soft labels" are required, such as when utilizing label smoothing or blended labels.
+
+    The `input` is expected to contain raw, unnormalized scores for each class and the `target`
+    should match the `input` shape and contain the target distribution over the classes.
+
+    The loss can be described as:
```
Comment (zou3519) [inline torch/nn/modules/loss.py:1136]: I got a bit confused as to why L is equal to l_{n, c}. I think what we meant was "The unreduced loss `L` is a 2D tensor such that each element `L_{n, c}` is equal to...

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -1121,6 +1125,91 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
+    .. math::
+        L = l_{n,c} = y_{n,c} \cdot \left( \log y_{n,c} - \\
+        \log\left(\frac{\exp(x_{n,c})}{\sum_{i=1}^C \exp(x_{n,i})} \right) \right)
+
+    where :math:`x_{n,c}` represents each element of ``input``, :math:`y_{n,c}` represents each element of
+    ``target``, and :math:`L` has the same shape as ``input`` and ``target``.  If :attr:`reduction` is not
+    ``'none'`` (default ``'mean'``), then:
+
+    .. math::
+        \ell(x, y) = \begin{cases}
```
Comment (zou3519) [inline torch/nn/modules/loss.py:1147]: I know the other losses have the same notation, but it's not clear that `l(x, y)` is the output of the module. We probably want something like "If reduction is not `none`, then the result is a reduced loss `l(x, y)` such that..."

File: aten/src/ATen/native/Loss.cpp [hunk 5]
```diff
@@ -123,6 +123,21 @@ Tensor kl_div_backward_cpu(const Tensor& grad, const Tensor& input, const Tensor
   return grad_input;
 }
 
+Tensor cross_entropy_loss_with_soft_labels(
+    const Tensor& self,
+    const Tensor& target,
+    int64_t reduction) {
+  TORCH_CHECK(self.dim() == 2, "Expected 2D input; got shape: ", self.sizes());
```
[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/native/Loss.cpp:130]: nit: Say the name of the function: "cross_entropy_loss_with_soft_labels(input, target): Expected 2d input..."

File: torch/csrc/api/include/torch/nn/modules/loss.h [hunk 6]
```diff
@@ -747,6 +747,45 @@ struct TORCH_API CrossEntropyLossImpl : public Cloneable<CrossEntropyLossImpl> {
 /// module storage semantics.
 TORCH_MODULE(CrossEntropyLoss);
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ CrossEntropyLossWithSoftLabels ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Creates a criterion that combines :func:`nn.LogSoftmax` and
+/// :func:`nn.KLDivLoss` in one single class.
```
Comment (zou3519) [inline torch/csrc/api/include/torch/nn/modules/loss.h:753]: nit: torch:nn::LogSoftmax and torch::nn::KLDivLoss (is that what they're called?)

File: test/cpp/api/functional.cpp [hunk 7]
```diff
@@ -843,6 +843,20 @@ TEST_F(FunctionalTest, CrossEntropy) {
...
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST_F(FunctionalTest, CrossEntropyWithSoftLabels) {
+  auto input = torch::tensor({{3., 3.}, {2., 2.}}, torch::kFloat);
+  auto target = torch::tensor({{1., 0.}, {0., 1.}}, torch::kFloat);
+  auto output = F::cross_entropy_with_soft_labels(
+      input, target, F::CrossEntropyWithSoftLabelsFuncOptions().reduction(torch::kMean));
+  // Using one-hot labels should be equivalent to cross entropy loss except
+  // smaller by a factor of num classes (2 in this case) for the mean reduction.
+  auto expected = torch::tensor(0.6931 / 2., torch::kFloat);
```
Comment (zou3519) [inline test/cpp/api/functional.cpp:854]: nit: It says "Using one-hot labels should be equivalent to cross entropy loss except" but then we hardcode in "torch::tensor(0.6931 / 2., torch::kFloat)", would it be better if we called the c++ version of F.cross_entropy and then used that as the `expected` tensor?

Ditto in the module test

File: test/cpp/api/modules.cpp [hunk 8]
```diff
@@ -4865,6 +4884,12 @@ TEST_F(ModulesTest, PrettyPrinCrossEntropyLoss) {
       c10::str(CrossEntropyLoss()), "torch::nn::CrossEntropyLoss()");
 }
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST_F(ModulesTest, PrettyPrinCrossEntropyLossWithSoftLabels) {
```
Comment (zou3519) [inline test/cpp/api/modules.cpp:4888]: nit: PrettyPrin -> PrettyPrint

File: test/cpp_api_parity/parity-tracker.md [hunk 9]
```diff
@@ -110,6 +110,7 @@ torch::nn::PairwiseDistance|Yes|No
 torch::nn::L1Loss|Yes|No
 torch::nn::MSELoss|Yes|No
 torch::nn::CrossEntropyLoss|Yes|No
+torch::nn::CrossEntropyLossWithSoftLabels|Yes|No
```
Comment (zou3519) [inline test/cpp_api_parity/parity-tracker.md:113]: It's been a long time since I've seen the parity tracker. Is there a comment or documentation somewhere about what tests this auto-generates?

File: torch/testing/_internal/common_nn.py [hunk 10]
```diff
@@ -3796,6 +3796,13 @@ def cross_entropy_loss_reference(input, target, weight=None, ignore_index=-100,
         reduction=reduction)
 
 
+def cross_entropy_loss_with_soft_labels_reference(input, target, reduction='mean'):
+    return kldivloss_reference(
+        torch.log_softmax(input, 1),
+        target,
+        reduction=reduction)
+
```
Comment (zou3519) [inline torch/testing/_internal/common_nn.py:3804]: Somewhere we should test that (1) cross_entropy_loss_with_soft_labels requires its input to be 2D, (2) cross_entropy_loss_with_soft_labels requires the input and target to be of the same shape and (3) the error messages for (1) and (2) are sane

File: torch/testing/_internal/common_nn.py [hunk 11]
```diff
@@ -4230,6 +4238,11 @@ def padding3d_circular(input, pad):
         target_fn=lambda: torch.empty(15).uniform_().mul(10).floor().long(),
         desc='weights',
     ),
+    dict(
+        module_name='CrossEntropyLossWithSoftLabels',
+        input_size=(15, 10),
+        target_fn=lambda: torch.empty(15, 10).uniform_(),
+    ),
```
Comment (zou3519) [inline torch/testing/_internal/common_nn.py:4245]: Does this need a reference_fn?

File: torch/testing/_internal/common_nn.py [hunk 12]
```diff
@@ -4523,6 +4536,16 @@ def padding3d_circular(input, pad):
...
+    dict(
+        module_name='CrossEntropyLossWithSoftLabels',
+        input_size=(5, 3),
+        target_fn=lambda: torch.rand(5, 3),
+        reference_fn=lambda i, t, m:
+            loss_reference_fns['CrossEntropyLossWithSoftLabels'](i, t, reduction=get_reduction(m)),
+        check_sum_reduction=True,
+        desc='2d',
+        check_bfloat16=False,
+    ),
```
Comment (zou3519) [inline torch/testing/_internal/common_nn.py:4548]: We should test all of the reduction options. Does this one only test `reduction='sum'`, or does it test all three? (sum, mean, none)

File: torch/nn/modules/loss.py (hunk 3 above)
Comment (zou3519) [reply torch/nn/modules/loss.py:1136]: https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html seems like a good example to follow

**Review by zou3519 — COMMENTED** [review]
I had some comments on the docs and the testing (I'm not sure what's being tested). Other than that, the implementations look correct to me

File: test/cpp/api/functional.cpp (hunk 7 above)
Comment (jbschlosser) [reply test/cpp/api/functional.cpp:854]: Cool, I like that better

File: torch/testing/_internal/common_nn.py (hunk 10 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_nn.py:3804]: Agreed, I'm going to add a set of tests into `test_nn.py` for now. Longer term, this sort of thing will likely be covered by tests generated from a `ModuleInfo` entry.

File: torch/nn/modules/loss.py (hunk 3 above)
Comment (jbschlosser) [reply torch/nn/modules/loss.py:1136]: > I got a bit confused as to why L is equal to l_{n, c}. I think what we meant was "The unreduced loss L is a 2D tensor such that each element L_{n, c} is equal to...

Good point, the L equal to l_{n, c} notation is arguably just wrong and should be changed- I like the phrasing you chose.

> https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html seems like a good example to follow

I think KLDivLoss is a nice and readable example, but it's hard to follow here - unlike for KLDivLoss, the loss here can't be directly defined element-wise because of the log-softmax.

File: aten/src/ATen/native/Loss.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/native/Loss.cpp:130]: From the [TF docs](https://www.tensorflow.org/api_docs/python/tf/keras/losses/CategoricalCrossentropy), it supports shape: `[batch_size, d0, .. dN]`. I'll remove the 2-dim constraint here as well since there's no reason for it (both components support > 2 dims).

File: torch/csrc/api/include/torch/nn/modules/loss.h (hunk 6 above)
Comment (jbschlosser) [reply torch/csrc/api/include/torch/nn/modules/loss.h:753]: Note that this is consistent with the C++ module `CrossEntropyLoss` docs, so I'll make an update there as well.

File: test/cpp/api/modules.cpp (hunk 8 above)
Comment (jbschlosser) [reply test/cpp/api/modules.cpp:4888]: Fixed for the original CrossEntropy as well

File: torch/testing/_internal/common_nn.py (hunk 11 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_nn.py:4245]: For these, I just copied & tweaked the corresponding CrossEntropy tests.

There's a `reference_fn` defined for the entry with `check_sum_reduction=True`, and that one does strictly more than this entry, so I'll just delete this one.

File: torch/testing/_internal/common_nn.py (hunk 12 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_nn.py:4548]: Unfortunately it looks like it only tests `reduction='sum'` :( (why this does exist? haha) I'll add some manual tests in `test_nn.py` for the other cases.

File: torch/testing/_internal/common_nn.py (hunk 10 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_nn.py:3804]: Agreed, the error cases will be covered in new tests within `test_nn.py`

File: test/cpp_api_parity/parity-tracker.md (hunk 9 above)
Comment (jbschlosser) [reply test/cpp_api_parity/parity-tracker.md:113]: No docs that I'm aware of :( My understanding is that it does this:

1. Instantiates a module in both C++ / python using the `constructor_args` / `cpp_constructor_args` entries in the `criterion_tests` dict entry corresponding to the module. If those aren't specified, it defaults to no args, which is okay for this loss.
2. It sends `input` / `target` through the C++ / python modules
3. It compares the resulting outputs

I noticed that the parity tracker checks functional forms as well, so I added that.

File: torch/nn/modules/loss.py [hunk 13]
```diff
@@ -1121,6 +1125,91 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
+    one single class. It is useful when training a classification problem with `C` classes when
+    "soft labels" are required, such as when utilizing label smoothing or blended labels.
+
+    The `input` is expected to contain raw, unnormalized scores for each class and the `target`
+    should match the `input` shape and contain the target distribution over the classes.
+
+    The loss can be described as:
+
+    .. math::
+        L = l_{n,c} = y_{n,c} \cdot \left( \log y_{n,c} - \\
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/loss.py]: Updated the docs to immediately define x and y before the equation is shown. If it's still unclear, I'll look into using `t` for the target, but I ideally wanted to keep it consistent with other docs

File: torch/nn/modules/loss.py (hunk 3 above)
Comment (jbschlosser) [reply torch/nn/modules/loss.py]: Okay, I decided to follow what `CrossEntropyLoss` does here: they just show the equation for softmax without explicitly defining the indices it is computed over

File: torch/nn/modules/loss.py [hunk 14]
```diff
@@ -1121,6 +1125,93 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
+        reduction (string, optional): Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will
+            be applied, ``'mean'``: the mean of the output is taken,
+            ``'sum'``: the output will be summed. Default: ``'mean'``
+
+    Shape:
+        - Input: :math:`(*, C)` where :math:`*` represents any number of dimensions (including none) and
+          `C = number of classes`
+        - Target: :math:`(*, C)`, same shape as the input
+        - Output: If :attr:`reduction` is ``'none'``, then :math:`(*, C)`, otherwise scalar
```
Comment (zou3519) [inline torch/nn/modules/loss.py:1176]: PyTorch CrossEntropyLoss accepts input of shape (N, C, *) and target of shape (N, *) where * is any number of dimensions. Do the shapes here make CrossEntropyLoss and CrossEntropyLossWithSoftLabels inconsistent?

Comment (zou3519) [reply torch/nn/modules/loss.py:1176]: In TF images are NHWC by default (instead of NCHW in PyTorch) so that's why they don't have this problem...

File: test/test_nn.py [hunk 15]
```diff
@@ -16387,6 +16389,50 @@ def cosine_distance(x, y):
             self.assertEqual(functional, modular, atol=1e-6, rtol=1e-6)
             self.assertEqual(traced, modular, atol=1e-6, rtol=1e-6)
 
+    @onlyOnCPUAndCUDA
+    def test_cross_entropy_loss_with_soft_labels(self, device):
+        N, C = 5, 3
+        input = torch.randn(N, C)
+        target = torch.tensor([1, 2, 0, 1, 2], dtype=torch.long)
```
Comment (zou3519) [inline test/test_nn.py:16396]: We should test correctness on an example with more than 2 dimensions

File: test/test_nn.py [hunk 16]
```diff
@@ -16387,6 +16389,50 @@ def cosine_distance(x, y):
             self.assertEqual(functional, modular, atol=1e-6, rtol=1e-6)
             self.assertEqual(traced, modular, atol=1e-6, rtol=1e-6)
 
+    @onlyOnCPUAndCUDA
+    def test_cross_entropy_loss_with_soft_labels(self, device):
```
Comment (zou3519) [inline test/test_nn.py:16393]: nit:  It would be nice to test correctness for some cases where `target` is not one-hot-encoded

Comment (jbschlosser) [reply test/test_nn.py:16393]: Note that the tests generated from the `criterion_tests` entry check against a reference_fn for the non-one-hot encoded target case.

**Review by zou3519 — COMMENTED** [review]
Implementation looks correct and some minor comments on the testing. I had two main concerns:
1. naming. Is it really OK for us to call this nn.CrossEntropyLossWithSoftLabels if it doesn't exactly compute "cross entropy"? Should we provide an option to actually cross_entropy and have that be the default? The implementation of that would not be very difficult (it's just `- (input *F.log_softmax(target)).sum()` and we could optimize that with some cpu/cuda kernels if necessary in the future
2. Multidimensional shape behavior: CrossEntropyLoss takes in inputs and targets of shape (N, C, *) and (N, *) respectively. To be consistent with that I would have expected us to accept inputs and targets of shape (N, C, *) and (N, C, *) but I might be missing something here

File: torch/nn/modules/loss.py (hunk 14 above)
Comment (jbschlosser) [reply torch/nn/modules/loss.py:1176]: You're right that they're inconsistent, but there's a few things to consider that made this choice hard:
* As you pointed out below, target shape == input shape for `CrossEntropyLossWithSoftLabels`, so that part can't be consistent with `CrossEntropyLoss`
* We don't want to unnecessarily bake in the need for `N`, since that will have to be undone for the no-batch-dim support work. To be consistent with `CrossEntropyLoss`, this would mean supporting shapes `(C)`, `(N, C)`, `(N, C, d1)`, `(N, C, d1, d2)`, etc. (i.e. sometimes `C` is in at dim 0 and sometimes it's at dim 1). I find this to be pretty confusing and honestly don't understand why the original CE did it this way, when `d1`, `d2`, etc. are semantically just more batch dimensions.

An interesting thing to consider could be supporting a configurable location for the `C` dim (i.e. the dim that log-softmax is taken over). But we couldn't easily make the default match CE's behavior, since the location of the `C` dim changes as the number of dims changes (location is 0 or -1 for shape `(C)`, 1 or -1 for shape `(N, C)`, and 1 for shapes `(N, C, d1, ...)`).

Also, admittedly confusingly, the `C` in `NHWC` / `NCHW` refers to "channels" (e.g. red, green, blue) while the `C` here refers to "number of classes".

**Comment by jbschlosser** [comment]
Closing since #61044 landed.
